### DISCUSSION TITLE
Exposing Keycloak Internally and Externally

### DISCUSSION THREAD

**Discussion by jseiser** [keycloak/keycloak#34712] [Q&A] [answered]
Hello,

We currently run keycloak in kubernetes, and expose it publicly under 1 hostname, with a WAF enforcing rules, and privately with different name.

e.g.

keycloak.domain
keycloak-int.domain  

We have end users, who can only access our systems through a DMZ, and they do not have any internet access. 

We configure Keycloak with TF, which hits KC at the public address, as well as some users who hit the system over the internet and also log into via that public KC.

When users coming through the DMZ's hit `keycloak-int.domain` it will fail because keycloak will always try to do a redirect to `keycloak.domain`, which they cant access, since they do not have internet access.

Is there any recommended way to run keycloak in this type of split brained setup?  Where we need to be able to handle both public access (TF configuration, user logon) as well as private access ( users who access through a dmz)

Thanks,

**Comment by embesozzi** [comment]
It's all about configuring the Keycloak hostname options. Generally speaking (because the latest version introduced some changes to the environment variables), you need to set hostname-strict to false. This usually indicates that you have a proxy in front of Keycloak, and it will use the `X-Forwarded-*` headers. The proxy must control these headers to allow your domains.

Alternatively, if you only want to expose the admin console internally, you can configure hostname-admin with a different DNS.

You can find a detailed explanation here [1].

[1] https://www.keycloak.org/server/hostname

**Reply by jseiser** [reply]
@embesozzi 

Assuming we can get ` hostname-strict to false`, the bitnami chart OOTB does some things that sets it to true, what are the odds this would work with Proxy Protocol instead of host headers?  We are fronting with an AWS NLB, which does not forward those headers.

Thanks.

**Reply by embesozzi** [reply, marked as answer]
Disclaimer: I haven't tried this option, but Keycloak supports the proxy-protocol-enabled flag [1].

[1] https://www.keycloak.org/server/reverseproxy
